File: backend-node/src/endpoints/refundEndpoints.ts
import { Router } from 'express';
import { RefundErrorCode, RefundService } from '../services/refundService.js';
import { refundToDto } from './dtos.js';

const STATUS_BY_CODE: Record<RefundErrorCode, number> = {
  IDEMPOTENCY_KEY_REQUIRED: 400,
  REASON_REQUIRED: 400,
  TRANSACTION_NOT_FOUND: 404,
  INVALID_STATUS: 409,
  REFUND_WINDOW_EXPIRED: 409
};

interface RefundRequestBody {
  reason?: string;
}

export function refundEndpoints(refundService: RefundService): Router {
  const router = Router();

  router.post('/api/transactions/:id/refunds', (req, res) => {
    const idempotencyKey = req.header('Idempotency-Key') ?? undefined;
    const body = req.body as RefundRequestBody;

    const result = refundService.refundFull(req.params.id, body?.reason, idempotencyKey);

    if (!result.ok) {
      res.status(STATUS_BY_CODE[result.code]).json({ code: result.code, message: result.message });
      return;
    }

    res.status(result.replay ? 200 : 201).json(refundToDto(result.refund));
  });

  return router;
}

File: backend/src/PaymentLab.Api/Endpoints/RefundEndpoints.cs
using PaymentLab.Api.Services;

namespace PaymentLab.Api.Endpoints;

public static class RefundEndpoints
{
    public static IEndpointRouteBuilder MapRefundEndpoints(this IEndpointRouteBuilder app)
    {
        // POST /api/transactions/{id}/refunds
        // Requires header: Idempotency-Key
        app.MapPost("/api/transactions/{id}/refunds", (
            string id,
            RefundRequest body,
            HttpRequest request,
            RefundService refunds) =>
        {
            var idempotencyKey = request.Headers["Idempotency-Key"].ToString();

            var result = refunds.RefundFull(id, body?.Reason ?? string.Empty, idempotencyKey);

            if (!result.Success)
            {
                var error = new ErrorDto(result.ErrorCode!, result.ErrorMessage!);
                return result.ErrorCode switch
                {
                    "TRANSACTION_NOT_FOUND" => Results.NotFound(error),
                    "TRANSACTION_NOT_REFUNDABLE" => Results.Conflict(error),
                    _ => Results.BadRequest(error)
                };
            }

            return Results.Ok(RefundDto.From(result.Refund!));
        });

        return app;
    }
}
